File: windows/src/Core/ExtensionInstaller.cs
#nullable enable
using System;
using System.Collections.Generic;
using System.IO;
using System.IO.Compression;
using System.Linq;
using System.Net.Http;
using System.Reflection;
using System.Security.Cryptography;
using System.Text;
using System.Text.RegularExpressions;
using System.Threading;
using System.Threading.Tasks;
using Newtonsoft.Json;
using YTray.Models;

namespace YTray.Core
{
    /// <summary>
    /// Downloads + installs the Yakit browser extension from the OSS release manifest
    /// (mirrors macOS ExtensionInstaller). The enterprise variant ships an unpacked
    /// extension whose zip root already contains manifest.json, so the extracted
    /// directory registers directly with InstanceStore.AddPlugin.
    /// </summary>
    public static class ExtensionInstaller
    {
        public const string EnterpriseVariant = "chrome-enterprise";
        // Identifies the managed plugin across upgrades; the extension's manifest name
        // ("Yakit Browser Agent") stays stable, so match on it instead of tracking paths.
        public const string ExtensionName = "Yakit Browser Agent";

        public static readonly Uri ManifestURL = new Uri("https://aliyun-oss.yaklang.com/chrome-extension/manifest.json");

        // The OSS mirror compresses responses for plain .NET Framework HttpClient requests
        // (no Accept-Encoding handling by default), so decode gzip/deflate explicitly.
        private static readonly HttpClient Http = CreateHttpClient();

        private sealed class BundledExtensionDescriptor
        {
            public string Version { get; set; } = "";
            public string Sha256 { get; set; } = "";
            public long Size { get; set; }
            public string Variant { get; set; } = "";
        }

        private static HttpClient CreateHttpClient()
        {
            var handler = new HttpClientHandler
            {
                AutomaticDecompression = System.Net.DecompressionMethods.GZip
                    | System.Net.DecompressionMethods.Deflate,
            };
            return new HttpClient(handler) { Timeout = TimeSpan.FromMinutes(5) };
        }

        public static string PluginsRoot(string applicationDirectory) =>
            Path.Combine(applicationDirectory, "Plugins");

        public static string PluginDirectory(string applicationDirectory, string version) =>
            Path.Combine(PluginsRoot(applicationDirectory), "yakit-browser-agent", version);

        public static string? ChromiumExtensionId(BrowserPlugin plugin)
        {
            if (plugin == null || string.IsNullOrWhiteSpace(plugin.Path)) return null;
            try
            {
                string? key = null;
                var manifestPath = Path.Combine(plugin.Path, "manifest.json");
                if (File.Exists(manifestPath))
                    key = JsonConvert.DeserializeObject<PluginManifest>(File.ReadAllText(manifestPath))?.Key;
                return ChromiumExtensionId(plugin.Path, key);
            }
            catch { return null; }
        }

        internal static string ChromiumExtensionId(string extensionPath, string? manifestKey)
        {
            var identity = string.IsNullOrWhiteSpace(manifestKey)
                ? Encoding.Unicode.GetBytes(Path.GetFullPath(extensionPath))
                : Convert.FromBase64String(manifestKey);
            using (var sha = SHA256.Create())
            {
                var digest = sha.ComputeHash(identity);
                var id = new StringBuilder(32);
                for (var i = 0; i < 16; i++)
                {
                    id.Append((char)('a' + (digest[i] >> 4)));
                    id.Append((char)('a' + (digest[i] & 15)));
                }
                return id.ToString();
            }
        }

        private static string ManagedExtensionOptOutPath(string applicationDirectory) =>
            Path.Combine(PluginsRoot(applicationDirectory), ".yakit-browser-agent-removed");

        public static bool TryGetBundledVersion(out string version)
        {
            version = "";
            try
            {
                var package = ReadBundledPackage(includeArchive: false);
                if (package.Descriptor == null) return false;
                version = package.Descriptor.Version;
                return !string.IsNullOrWhiteSpace(version);
            }
            catch { return false; }
        }

        /// <summary>
        /// Extracts the release archive embedded by the packaging workflow. The archive is
        /// revalidated at runtime and expanded beneath the normal per-user Plugins directory.
        /// </summary>
        public static bool TryInstallBundled(string applicationDirectory, out string directory,
            out string version, bool ignoreOptOut = false, bool replaceExisting = false)
        {
            directory = "";
            version = "";
            if (!ignoreOptOut && File.Exists(ManagedExtensionOptOutPath(applicationDirectory))) return false;

            var package = ReadBundledPackage(includeArchive: true);
            if (package.Descriptor == null || package.Archive == null) return false;
            version = package.Descriptor.Version;
            ValidateBundledPackage(package.Descriptor, package.Archive);

            var destination = PluginDirectory(applicationDirectory, version);
            try
            {
                if (!replaceExisting)
                {
                    directory = ResolveExtensionRoot(destination);
                    return true;
                }
            }
            catch { }

            var temporary = destination + ".partial-" + Guid.NewGuid().ToString("N");
            try
            {
                ExtractArchiveSafely(package.Archive, temporary);
                _ = ResolveExtensionRoot(temporary);
                if (Directory.Exists(destination)) Directory.Delete(destination, true);
                Directory.CreateDirectory(Path.GetDirectoryName(destination)!);
                Directory.Move(temporary, destination);
                directory = ResolveExtensionRoot(destination);
                return true;
            }
            finally
            {
                try { if (Directory.Exists(temporary)) Directory.Delete(temporary, true); } catch { }
            }
        }

        public static void MarkManagedExtensionRemoved(string applicationDirectory)
        {
            try
            {
                Directory.CreateDirectory(PluginsRoot(applicationDirectory));
                File.WriteAllText(ManagedExtensionOptOutPath(applicationDirectory), DateTime.UtcNow.ToString("O"));
            }
            catch { }
        }

        public static void ClearManagedExtensionRemoved(string applicationDirectory)
        {
            try
            {
                var marker = ManagedExtensionOptOutPath(applicationDirectory);
                if (File.Exists(marker)) File.Delete(marker);
            }
            catch { }
        }

        private static (BundledExtensionDescriptor? Descriptor, byte[]? Archive) ReadBundledPackage(bool includeArchive)
        {
            var assembly = Assembly.GetExecutingAssembly();
            var names = assembly.GetManifestResourceNames();
            var descriptorName = names.FirstOrDefault(name =>
                name.EndsWith("BundledExtension.bundled-extension.json", StringComparison.OrdinalIgnoreCase));
            if (descriptorName == null) return (null, null);

            BundledExtensionDescriptor? descriptor;
            using (var stream = assembly.GetManifestResourceStream(descriptorName))
            using (var reader = stream == null ? null : new StreamReader(stream, Encoding.UTF8, true))
                descriptor = reader == null ? null : JsonConvert.DeserializeObject<BundledExtensionDescriptor>(reader.ReadToEnd());
            if (!includeArchive || descriptor == null) return (descriptor, null);

            var archiveName = names.FirstOrDefault(name =>
                name.EndsWith("BundledExtension.yakit-browser-agent.zip", StringComparison.OrdinalIgnoreCase));
            if (archiveName == null) return (descriptor, null);
            using (var stream = assembly.GetManifestResourceStream(archiveName))
            using (var memory = new MemoryStream())
            {
                if (stream == null) return (descriptor, null);
                stream.CopyTo(memory);
                return (descriptor, memory.ToArray());
            }
        }

        private static void ValidateBundledPackage(BundledExtensionDescriptor descriptor, byte[] archive)
        {
            if (!Regex.IsMatch(descriptor.Version ?? "", @"^[0-9]+(?:\.[0-9]+)*$", RegexOptions.CultureInvariant)
                || !string.Equals(descriptor.Variant, EnterpriseVariant, StringComparison.OrdinalIgnoreCase))
                throw new YTrayException(YTrayError.ExtensionInstallFailed, "内置插件元数据无效");
            if (archive.LongLength != descriptor.Size)
                throw new YTrayException(YTrayError.ExtensionInstallFailed, "内置插件大小校验失败");
            using (var sha = SHA256.Create())
            {
                var actual = string.Concat(sha.ComputeHash(archive).Select(value => value.ToString("x2")));
                if (!string.Equals(actual, descriptor.Sha256, StringComparison.OrdinalIgnoreCase))
                    throw new YTrayException(YTrayError.ExtensionInstallFailed, "内置插件 SHA-256 校验失败");
            }
        }

        private static void ExtractArchiveSafely(byte[] archive, string destination)
        {
            Directory.CreateDirectory(destination);
            var destinationRoot = Path.GetFullPath(destination)
                .TrimEnd(Path.DirectorySeparatorChar, Path.AltDirectorySeparatorChar)
                + Path.DirectorySeparatorChar;
            using (var memory = new MemoryStream(archive, writable: false))
            using (var zip = new ZipArchive(memory, ZipArchiveMode.Read))
            {
                foreach (var entry in zip.Entries)
                {
                    var relative = entry.FullName.Replace('/', Path.DirectorySeparatorChar);
                    var output = Path.GetFullPath(Path.Combine(destination, relative));
                    if (!output.StartsWith(destinationRoot, StringComparison.OrdinalIgnoreCase))
                        throw new YTrayException(YTrayError.ExtensionInstallFailed, "内置插件包含不安全路径");
                    if (string.IsNullOrEmpty(entry.Name))
                    {
                        Directory.CreateDirectory(output);
                        continue;
                    }
                    Directory.CreateDirectory(Path.GetDirectoryName(output)!);
                    using (var input = entry.Open())
                    using (var file = new FileStream(output, FileMode.Create, FileAccess.Write, FileShare.None))
                        input.CopyTo(file);
                }
            }
        }

        public static async Task<ExtensionManifest> FetchManifestAsync()
        {
            using (var resp = await Http.GetAsync(ManifestURL))
            {
                if (!resp.IsSuccessStatusCode) throw new YTrayException(YTrayError.ExtensionInstallFailed, "插件清单返回异常");
                var json = await resp.Content.ReadAsStringAsync();
                var manifest = JsonConvert.DeserializeObject<ExtensionManifest>(json);
                if (manifest == null || string.IsNullOrWhiteSpace(manifest.Latest))
                    throw new YTrayException(YTrayError.ExtensionInstallFailed, "插件清单格式无效");
                return manifest;
            }
        }

        public static ExtensionArtifact? EnterpriseArtifact(ExtensionReleaseVersion version) =>
            version.Artifacts.FirstOrDefault(a =>
                string.Equals(a.Variant, EnterpriseVariant, StringComparison.OrdinalIgnoreCase)
                && !string.IsNullOrWhiteSpace(a.Url));

        /// <summary>
        /// Compares dotted numeric versions ("0.2.10" &gt; "0.2.2"). Non-numeric segments
        /// fall back to ordinal comparison; a parse failure never throws.
        /// </summary>
        public static int CompareVersions(string? left, string? right)
        {
            var l = (left ?? "").Trim().TrimStart('v', 'V');
            var r = (right ?? "").Trim().TrimStart('v', 'V');
            var leftParts = l.Split('.');
            var rightParts = r.Split('.');
            var count = Math.Max(leftParts.Length, rightParts.Length);
            for (var i = 0; i < count; i++)
            {
                var lp = i < leftParts.Length ? leftParts[i] : "";
                var rp = i < rightParts.Length ? rightParts[i] : "";
                if (long.TryParse(lp, out var ln) && long.TryParse(rp, out var rn))
                {
                    if (ln != rn) return ln.CompareTo(rn);
                }
                else
                {
                    var ordinal = string.CompareOrdinal(lp, rp);
                    if (ordinal != 0) return ordinal > 0 ? 1 : -1;
                }
            }
            return string.CompareOrdinal(l, r);
        }

        /// <summary>
        /// Returns true only when an embedded plugin should be installed automatically.
        /// A missing local plugin is initialized from the bundle; an existing plugin is
        /// upgraded only when the embedded version is strictly newer. Equal versions are
        /// left untouched and a newer local version is never downgraded.
        /// </summary>
        public static bool ShouldInstallBundledVersion(string? bundledVersion, string? installedVersion,
            bool allowSameVersion = false)
        {
            if (string.IsNullOrWhiteSpace(bundledVersion)) return false;
            if (string.IsNullOrWhiteSpace(installedVersion)) return true;
            var comparison = CompareVersions(bundledVersion, installedVersion);
            return comparison > 0 || (allowSameVersion && comparison == 0);
        }

        /// <summary>
        /// Installs the given release into Plugins/yakit-browser-agent/{version} and returns
        /// the extracted directory. Existing installs of the same version are reinstalled
        /// (directory replaced), which doubles as a repair path.
        /// </summary>
        public static async Task<string> InstallAsync(ExtensionReleaseVersion version, string applicationDirectory,
            IProgress<RuntimeInstaller.InstallProgress>? progress = null, CancellationToken cancellationToken = default)
        {
            if (version == null) throw new ArgumentNullException(nameof(version));
            var artifact = EnterpriseArtifact(version)
                ?? throw new YTrayException(YTrayError.ExtensionInstallFailed, $"版本 {version.Version} 没有 {EnterpriseVariant} 产物");
            var tmpZip = Path.Combine(Path.GetTempPath(), "ytray-ext-" + Guid.NewGuid() + ".zip");
            try
            {
                progress?.Report(new RuntimeInstaller.InstallProgress { Percent = 2, Message = "正在连接下载镜像…" });
                await RuntimeInstaller.DownloadAsync(Http, new Uri(artifact.Url), tmpZip, artifact.Size, progress, cancellationToken);
                cancellationToken.ThrowIfCancellationRequested();
                progress?.Report(new RuntimeInstaller.InstallProgress { Percent = 81, Message = "正在校验 SHA-256…" });
                var actualHash = await RuntimeInstaller.HashFileAsync(tmpZip);
                if (!string.IsNullOrEmpty(artifact.Sha256)
                    && !string.Equals(actualHash, artifact.Sha256, StringComparison.OrdinalIgnoreCase))
                    throw new YTrayException(YTrayError.ExtensionInstallFailed, "SHA-256 校验失败");

                var dest = PluginDirectory(applicationDirectory, version.Version);
                if (Directory.Exists(dest)) try { Directory.Delete(dest, true); } catch { }
                Directory.CreateDirectory(dest);
                progress?.Report(new RuntimeInstaller.InstallProgress { Percent = 84, Message = "正在解压插件…" });
                await RuntimeInstaller.ExtractAsync(tmpZip, dest, progress, cancellationToken, YTrayError.ExtensionInstallFailed);

                if (!File.Exists(Path.Combine(dest, "manifest.json")))
                    throw new YTrayException(YTrayError.ExtensionInstallFailed, "ZIP 根目录缺少 manifest.json");
                return ResolveExtensionRoot(dest);
            }
            finally
            {
                try { if (File.Exists(tmpZip)) File.Delete(tmpZip); } catch { }
            }
        }

        private static string ResolveExtensionRoot(string directory)
        {
            if (File.Exists(Path.Combine(directory, "manifest.json"))) return directory;
            var nested = Directory.EnumerateDirectories(directory).ToList();
            if (nested.Count == 1 && File.Exists(Path.Combine(nested[0], "manifest.json")))
                return nested[0];
            throw new YTrayException(YTrayError.ExtensionInstallFailed, "解压后未找到 manifest.json");
        }

        /// <summary>
        /// Drops version directories of the managed plugin other than the installed one so
        /// upgrades do not accumulate old copies. Best-effort; locked files are ignored.
        /// </summary>
        public static void CleanupOldVersions(string applicationDirectory, string installedVersion)
        {
            try
            {
                var root = Path.Combine(PluginsRoot(applicationDirectory), "yakit-browser-agent");
                if (!Directory.Exists(root)) return;
                foreach (var dir in Directory.EnumerateDirectories(root))
                    if (!string.Equals(Path.GetFileName(dir), installedVersion, StringComparison.OrdinalIgnoreCase))
                        try { Directory.Delete(dir, true); } catch { }
            }
            catch { }
        }
    }
}

File: darwin/Sources/YTray/ExtensionInstaller.swift
import Foundation
import CryptoKit

enum ExtensionInstaller {
    static let enterpriseVariant = "chrome-enterprise"
    // Identifies the managed plugin across upgrades; the extension's manifest name
    // ("Yakit Browser Agent") stays stable, so match on it instead of tracking paths.
    static let extensionName = "Yakit Browser Agent"

    static let manifestURL = URL(string: "https://aliyun-oss.yaklang.com/chrome-extension/manifest.json")!

    private struct BundledExtensionDescriptor: Decodable {
        let version: String
        let sha256: String
        let size: Int64
        let variant: String
    }

    static func pluginsRoot(applicationDirectory: URL) -> URL {
        applicationDirectory.appendingPathComponent("Plugins", isDirectory: true)
    }

    static func pluginDirectory(applicationDirectory: URL, version: String) -> URL {
        pluginsRoot(applicationDirectory: applicationDirectory)
            .appendingPathComponent("yakit-browser-agent/\(version)", isDirectory: true)
    }

    private static func managedExtensionOptOutURL(applicationDirectory: URL) -> URL {
        pluginsRoot(applicationDirectory: applicationDirectory)
            .appendingPathComponent(".yakit-browser-agent-removed")
    }

    static func bundledVersion(resourcesURL: URL? = Bundle.main.resourceURL) -> String? {
        guard let package = try? bundledPackage(resourcesURL: resourcesURL) else { return nil }
        return package.descriptor.version
    }

    /// Installs the archive placed in YTray.app by the packaging workflow. The pinned archive
    /// is validated again at runtime before it is expanded into the user's Plugins directory.
    static func installBundled(
        into applicationDirectory: URL,
        resourcesURL: URL? = Bundle.main.resourceURL,
        ignoreOptOut: Bool = false,
        replaceExisting: Bool = false
    ) throws -> (directory: URL, version: String)? {
        if !ignoreOptOut,
           FileManager.default.fileExists(atPath: managedExtensionOptOutURL(applicationDirectory: applicationDirectory).path) {
            return nil
        }
        guard let package = try bundledPackage(resourcesURL: resourcesURL) else { return nil }
        let descriptor = package.descriptor
        let versionParts = descriptor.version.split(separator: ".", omittingEmptySubsequences: false)
        guard descriptor.variant.caseInsensitiveCompare(enterpriseVariant) == .orderedSame,
              !versionParts.isEmpty,
              versionParts.allSatisfy({ part in
                  !part.isEmpty && part.utf8.allSatisfy({ byte in byte >= 48 && byte <= 57 })
              }) else {
            throw YTrayError.extensionInstallFailed("内置插件元数据无效")
        }
        let attributes = try FileManager.default.attributesOfItem(atPath: package.archive.path)
        guard (attributes[.size] as? NSNumber)?.int64Value == descriptor.size else {
            throw YTrayError.extensionInstallFailed("内置插件大小校验失败")
        }
        guard try sha256(of: package.archive).caseInsensitiveCompare(descriptor.sha256) == .orderedSame else {
            throw YTrayError.extensionInstallFailed("内置插件 SHA-256 校验失败")
        }
        try validateArchiveEntries(package.archive)

        let destination = pluginDirectory(applicationDirectory: applicationDirectory, version: descriptor.version)
        if !replaceExisting, let existing = try? resolveExtensionRoot(destination) {
            return (existing, descriptor.version)
        }

        let temporary = destination.deletingLastPathComponent()
            .appendingPathComponent(destination.lastPathComponent + ".partial-" + UUID().uuidString, isDirectory: true)
        try? FileManager.default.removeItem(at: temporary)
        try FileManager.default.createDirectory(at: temporary, withIntermediateDirectories: true)
        defer { try? FileManager.default.removeItem(at: temporary) }

        let unzip = Process()
        unzip.executableURL = URL(fileURLWithPath: "/usr/bin/ditto")
        unzip.arguments = ["-x", "-k", package.archive.path, temporary.path]
        let errorPipe = Pipe()
        unzip.standardError = errorPipe
        try unzip.run()
        unzip.waitUntilExit()
        guard unzip.terminationStatus == 0 else {
            let data = errorPipe.fileHandleForReading.readDataToEndOfFile()
            throw YTrayError.extensionInstallFailed(String(data: data, encoding: .utf8) ?? "内置插件解压失败")
        }
        _ = try resolveExtensionRoot(temporary)
        try? FileManager.default.removeItem(at: destination)
        try FileManager.default.createDirectory(at: destination.deletingLastPathComponent(), withIntermediateDirectories: true)
        try FileManager.default.moveItem(at: temporary, to: destination)
        return (try resolveExtensionRoot(destination), descriptor.version)
    }

    static func markManagedExtensionRemoved(applicationDirectory: URL) {
        let marker = managedExtensionOptOutURL(applicationDirectory: applicationDirectory)
        try? FileManager.default.createDirectory(at: marker.deletingLastPathComponent(), withIntermediateDirectories: true)
        try? Data(Date().ISO8601Format().utf8).write(to: marker, options: .atomic)
    }

    static func clearManagedExtensionRemoved(applicationDirectory: URL) {
        try? FileManager.default.removeItem(at: managedExtensionOptOutURL(applicationDirectory: applicationDirectory))
    }

    private static func bundledPackage(resourcesURL: URL?) throws
        -> (descriptor: BundledExtensionDescriptor, archive: URL)? {
        guard let root = resourcesURL?.appendingPathComponent("BundledExtension", isDirectory: true) else { return nil }
        let descriptorURL = root.appendingPathComponent("bundled-extension.json")
        let archiveURL = root.appendingPathComponent("yakit-browser-agent.zip")
        guard FileManager.default.fileExists(atPath: descriptorURL.path),
              FileManager.default.fileExists(atPath: archiveURL.path) else { return nil }
        return (try JSONDecoder().decode(BundledExtensionDescriptor.self, from: Data(contentsOf: descriptorURL)), archiveURL)
    }

    private static func validateArchiveEntries(_ archive: URL) throws {
        let process = Process()
        process.executableURL = URL(fileURLWithPath: "/usr/bin/unzip")
        process.arguments = ["-Z1", archive.path]
        let output = Pipe()
        process.standardOutput = output
        process.standardError = Pipe()
        try process.run()
        let listingData = output.fileHandleForReading.readDataToEndOfFile()
        process.waitUntilExit()
        guard process.terminationStatus == 0,
              let listing = String(data: listingData, encoding: .utf8) else {
            throw YTrayError.extensionInstallFailed("无法检查内置插件压缩包")
        }
        for entry in listing.split(separator: "\n").map(String.init) {
            let components = entry.replacingOccurrences(of: "\\", with: "/").split(separator: "/")
            if entry.hasPrefix("/") || components.contains(where: { $0 == ".." }) {
                throw YTrayError.extensionInstallFailed("内置插件包含不安全路径")
            }
        }
    }

    static func fetchManifest() async throws -> ExtensionManifest {
        let (data, response) = try await URLSession.shared.data(from: manifestURL)
        guard (response as? HTTPURLResponse)?.statusCode == 200 else {
            throw YTrayError.extensionInstallFailed("插件清单返回异常")
        }
        do {
            let manifest = try JSONDecoder().decode(ExtensionManifest.self, from: data)
            guard !manifest.latest.isEmpty else {
                throw YTrayError.extensionInstallFailed("插件清单格式无效")
            }
            return manifest
        } catch let error as YTrayError {
            throw error
        } catch {
            throw YTrayError.extensionInstallFailed("插件清单格式无效")
        }
    }

    static func enterpriseArtifact(of version: ExtensionReleaseVersion) -> ExtensionArtifact? {
        version.artifacts.first { $0.variant.caseInsensitiveCompare(enterpriseVariant) == .orderedSame && !$0.url.isEmpty }
    }

    /// Compares dotted numeric versions ("0.2.10" > "0.2.2"). Non-numeric segments
    /// fall back to ordinal comparison; a parse failure never throws.
    static func compareVersions(_ left: String?, _ right: String?) -> ComparisonResult {
        func segments(_ raw: String?) -> [String] {
            (raw ?? "")
                .trimmingCharacters(in: .whitespaces)
                .drop { $0 == "v" || $0 == "V" }
                .split(separator: ".", omittingEmptySubsequences: false)
                .map(String.init)
        }
        let leftParts = segments(left)
        let rightParts = segments(right)
        for index in 0 ..< max(leftParts.count, rightParts.count) {
            let lp = index < leftParts.count ? leftParts[index] : ""
            let rp = index < rightParts.count ? rightParts[index] : ""
            if let ln = Int64(lp), let rn = Int64(rp) {
                if ln != rn { return ln < rn ? .orderedAscending : .orderedDescending }
            } else {
                let ordinal = lp.compare(rp, options: [.literal])
                if ordinal != .orderedSame { return ordinal }
            }
        }
        return (left ?? "").compare(right ?? "", options: [.literal])
    }

    /// Decides whether the embedded plugin should be installed automatically. A missing
    /// local plugin is initialized from the bundle; an existing plugin is upgraded only
    /// when the embedded version is strictly newer. This prevents equal-version rewrites
    /// and downgrades after a user has independently updated the plugin.
    static func shouldInstallBundledVersion(
        _ bundledVersion: String?,
        installedVersion: String?,
        allowSameVersion: Bool = false
    ) -> Bool {
        guard let bundledVersion,
              !bundledVersion.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else { return false }
        guard let installedVersion,
              !installedVersion.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else { return true }
        let comparison = compareVersions(bundledVersion, installedVersion)
        return comparison == .orderedDescending || (allowSameVersion && comparison == .orderedSame)
    }

    /// Installs the given release into Plugins/yakit-browser-agent/{version} and returns
    /// the extracted extension directory (the one containing manifest.json). Existing
    /// installs of the same version are replaced, which doubles as a repair path.
    static func install(version: ExtensionReleaseVersion, into applicationDirectory: URL) async throws -> URL {
        guard let artifact = enterpriseArtifact(of: version), let url = URL(string: artifact.url) else {
            throw YTrayError.extensionInstallFailed("版本 \(version.version) 没有 \(enterpriseVariant) 产物")
        }
        let (temporary, response) = try await URLSession.shared.download(from: url)
        guard (response as? HTTPURLResponse)?.statusCode == 200 else {
            throw YTrayError.extensionInstallFailed("ZIP 下载返回异常")
        }
        if !artifact.sha256.isEmpty {
            let actualHash = try sha256(of: temporary)
            guard actualHash.caseInsensitiveCompare(artifact.sha256) == .orderedSame else {
                throw YTrayError.extensionInstallFailed("SHA-256 校验失败")
            }
        }

        let destination = pluginDirectory(applicationDirectory: applicationDirectory, version: version.version)
        try? FileManager.default.removeItem(at: destination)
        try FileManager.default.createDirectory(at: destination, withIntermediateDirectories: true)
        let unzip = Process()
        unzip.executableURL = URL(fileURLWithPath: "/usr/bin/ditto")
        unzip.arguments = ["-x", "-k", temporary.path, destination.path]
        let errorPipe = Pipe()
        unzip.standardError = errorPipe
        try unzip.run()
        unzip.waitUntilExit()
        guard unzip.terminationStatus == 0 else {
            let data = errorPipe.fileHandleForReading.readDataToEndOfFile()
            throw YTrayError.extensionInstallFailed(String(data: data, encoding: .utf8) ?? "ZIP 解压失败")
        }
        return try resolveExtensionRoot(destination)
    }

    /// Registers the zip root directly; a single nested directory containing the manifest
    /// (defensive against future packaging changes) is also accepted.
    private static func resolveExtensionRoot(_ directory: URL) throws -> URL {
        if FileManager.default.fileExists(atPath: directory.appendingPathComponent("manifest.json").path) {
            return directory
        }
        let children = (try? FileManager.default.contentsOfDirectory(
            at: directory, includingPropertiesForKeys: nil, options: [.skipsHiddenFiles])) ?? []
        if children.count == 1,
           FileManager.default.fileExists(atPath: children[0].appendingPathComponent("manifest.json").path) {
            return children[0]
        }
        throw YTrayError.extensionInstallFailed("解压后未找到 manifest.json")
    }

    /// Drops version directories of the managed plugin other than the installed one so
    /// upgrades do not accumulate old copies. Best-effort; locked files are ignored.
    static func cleanupOldVersions(applicationDirectory: URL, installedVersion: String) {
        let root = pluginsRoot(applicationDirectory: applicationDirectory)
            .appendingPathComponent("yakit-browser-agent", isDirectory: true)
        let children = (try? FileManager.default.contentsOfDirectory(
            at: root, includingPropertiesForKeys: nil, options: [.skipsHiddenFiles])) ?? []
        for child in children where child.lastPathComponent != installedVersion {
            try? FileManager.default.removeItem(at: child)
        }
    }

    private static func sha256(of url: URL) throws -> String {
        let handle = try FileHandle(forReadingFrom: url)
        defer { try? handle.close() }
        var hasher = SHA256()
        while true {
            guard let chunk = try handle.read(upToCount: 1024 * 1024), !chunk.isEmpty else { break }
            hasher.update(data: chunk)
        }
        return hasher.finalize().map { String(format: "%02x", $0) }.joined()
    }
}
